lib/Reader/readerir.cpp

```diff
@@ -4070,19 +4084,19 @@ void GenIR::storeElem(ReaderBaseNS::StElemOpcode Opcode,
   }
 }
 
-// Get array element type.
-Type *GenIR::getArrayElementType(IRNode *Array,
-                                 CORINFO_RESOLVED_TOKEN *ResolvedToken,
-                                 CorInfoType *CorType,
-                                 ReaderAlignType *Alignment) {
+// Get array element type from the token and/or CorType.
+Type *GenIR::getMSILArrayElementType(IRNode *Array,
+                                     CORINFO_RESOLVED_TOKEN *ResolvedToken,
+                                     CorInfoType *CorType,
+                                     ReaderAlignType *Alignment) {
   ASSERTNR(Alignment != nullptr);
   ASSERTNR(CorType != nullptr);
   CORINFO_CLASS_HANDLE ClassHandle = nullptr;
   if (*CorType == CorInfoType::CORINFO_TYPE_CLASS) {
     PointerType *Ty = cast<PointerType>(Array->getType());
-    if (!isArrayType(Ty)) {
+    if (!isArrayType(Ty, nullptr)) {
       // Likely System.__Canon*. Assume array of object.
-      Ty = getArrayOfReferenceType();
+      Ty = getArrayOfElementType(this->getBuiltInObjectType());
     }
     StructType *ReferentTy = cast<StructType>(Ty->getPointerElementType());
```

Comment:
No. ensureIsArray mutates an array value. But this can be simplified. We just want the element type, so there is no need to actually construct the array of object just to then deduce that the element type is object. I'll simplify the code.